OSDOCS-7104: new update duration diagram (#63023)
openshift/openshift-docs: modules/update-duration-example.adoc

// Module included in the following assemblies:
//
// * updating/understanding_updates/understanding-openshift-update-duration.adoc

:_content-type: CONCEPT
[id="update-duration-example_{context}"]
= Example update duration of cluster Operators

image::update-duration.png[A diagram displaying the periods during which cluster Operators update themselves during an OCP update]

The previous diagram shows an example of the time that cluster Operators might take to update to their new versions.

> The example is based on a three-node AWS OVN cluster, which has a healthy compute `MachineConfigPool` and no workloads that take long to drain, updating from 4.13.0 to 4.14.0.

Reviewer: [The run we timelined][1] was [4.13.0-0.ci-2023-07-31-121322 to 4.14.0-0.ci-2023-07-31-212944][2].  Can we say "from 4.13 to 4.14" here, without pinning specific patch releases?

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.14-upgrade-from-stable-4.13-e2e-aws-ovn-upgrade/1686131647159734272
[2]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.14-upgrade-from-stable-4.13-e2e-aws-ovn-upgrade/1686131647159734272/artifacts/e2e-aws-ovn-upgrade/gather-extra/artifacts/clusterversion.json

Author: Sure thing, I mentioned the z stream in anticipation of you or someone else wanting to be more specific. But I have no issue limiting version references to the minor level.

[NOTE]
====

> The specific update duration of a cluster and its Operators can vary based on several cluster characteristics, such as the target version, the amount of nodes, and the types of workloads scheduled to the nodes.

Reviewer: > Did you notice this doc doesn't just quote you a number?  Yeah, we're  about to explain a bunch of dynamic factors below :)

We also have "OpenShift Container Platform update duration varies based on the deployment topology" at the top of the doc.  But I agree that it's worth reminding folks a few times in this doc that the durations are variable based on multiple factors, so I'm just calling out your note as both somewhat redundant, and still useful anyway :).

Author: Lol fair, I wanted some caveat near the image that technically introduced some concrete-ish numbers. If you're cool with keeping it I'd like to keep it, but if other reviewers call out that it's overly redundant as well, I can remove it.

Reviewer: I'm fine keeping it.  I agree that a local disclaimer near the image makes sense, even though we have similar disclaimers in other places.

====

Each cluster Operator has characteristics that affect the time it takes to update itself.
For instance, the Kube API Server Operator in this example took over eleven minutes to update because it was serving customer requests and is designed to be conservative about draining out existing requests before shutting down a pod.

> In the case of this Operator, update speed is sacrificed to help prevent and limit disruptions to cluster functionality during an update.

Author: I could go either way in terms of placement, both options have their merits.

> In the case of this Operator, update speed is sacrificed to help prevent and limit disruptions to cluster functionality during an update.

Author: Alright, moving this module to be under the two general discussion sections as I don't have a strong preference to keep it where it is now.

> In the case of this Operator, update speed is sacrificed to help prevent and limit disruptions to cluster functionality during an update.

Reviewer 2: the Kube API Server Operator in this example took over eleven minutes to update because kube-apiserver provides graceful termination support. For existing requests in-flight are allowed to complete gracefully. This may result in a longer shutdown of the kube-apiserver.

// Module included in the following assemblies:
//
// * updating/understanding_updates/understanding-openshift-update-duration.adoc

:_content-type: CONCEPT
[id="update-duration-example_{context}"]
= Example update duration of cluster Operators

> image::update-duration.png[A diagram displaying the periods during which cluster Operators update themselves during an OCP update]

Reviewer 3: Maybe put the image after the introductory paragraph so that folks know what the image is?

Author: I could see how that would give some context to the image, but it would also make users read about specific references to parts of the image in these following paragraphs without having the chance to see the image itself.

I think the section heading is at least enough for users to know what they're looking at before getting to the textual description, and would want to keep the image where it is, but let me know if you feel strongly about the image placement.

The previous diagram shows an example of the time that cluster Operators might take to update to their new versions.
The example is based on a three-node AWS OVN cluster, which has a healthy compute `MachineConfigPool` and no workloads that take long to drain, updating from 4.13.0 to 4.14.0.

[NOTE]
====
The specific update duration of a cluster and its Operators can vary based on several cluster characteristics, such as the target version, the amount of nodes, and the types of workloads scheduled to the nodes.
====

Each cluster Operator has characteristics that affect the time it takes to update itself.

> For instance, the Kube API Server Operator in this example took over eleven minutes to update because `kube-apiserver` provides graceful termination support, meaning that existing, in-flight requests are allowed to complete gracefully.

Reviewer 3: ```suggestion
For instance, the Kube API Server Operator in this example took more than eleven minutes to update because `kube-apiserver` provides graceful termination support, meaning that existing, in-flight requests are allowed to complete gracefully.
```

Reviewer 3: >  over eleven minutes

Is this a particularly long period of time to update?

> This may result in a longer shutdown of the `kube-apiserver`.

Reviewer 3: ```suggestion
This might result in a longer shutdown of the `kube-apiserver`.
```

In the case of this Operator, update speed is sacrificed to help prevent and limit disruptions to cluster functionality during an update.

> Another characteristic that affects the update duration of an Operator is whether the Operator utilizes DaemonSets.

Reviewer 3: This phrase applies to daemon sets for all Operators, yes?

```suggestion
Another characteristic that affects the update duration of an Operator is whether the Operator utilizes daemon sets, which can take time to roll out their version changes.
```

Author: @wking could confirm this. I'm personally not sure whether it applies to all Operators that utilize daemon sets, or only Operators that utilize "full-cluster daemon sets" (or whether that distinction makes sense, I really don't know anything about daemon sets haha)

Reviewer: I agree with Seb.  There are some OpenShift DaemonSets that limit themselves to just control-plane nodes, like `openshift-machine-config-operator`'s `machine-config-server`.  The number of compute nodes does not affect the duration of updating that DaemonSet.  The machine-config operator also has a `machine-config-daemon` DaemonSet, which is full-cluster, and which updates with `maxUnavailable: 10%`.  But the time it takes to roll that DaemonSet is much less than the time it takes to roll the MachineConfigPools, so we don't usually bother calling out the `machine-config-daemon` roll time.

Author: Given that information, I think it might make sense to not swap the phrasing and mention only daemon sets in general here, while keeping "full-cluster DaemonSets, which can take time to roll out their version changes," down below (so to be more specific about how the full-cluster daemon sets in particular add a lot of time to an Operator's update duration).

@mburke5678 let me know if you still want me to swap the phrasing and make it work.

> The Network and DNS Operators utilize full-cluster DaemonSets, which can take time to roll out their version changes, and this is one of several reasons why these Operators might take longer to update themselves.

Reviewer 3: ```suggestion
The Network and DNS Operators utilize full-cluster daemon sets, which is one of several reasons why these Operators might take longer to update themselves.
```

Author: Ditto about keeping this phrasing here, as per my reply to your comment above.
